src/new.tsx

```diff
@@ -93,16 +99,86 @@ function Charming() {
     }
   }, [data, points]);
 
+  React.useEffect(() => {
+    // Data is required to determine grapheme cluster breaks in pasted strings
+    if (!data) return;
+
+    const onKeyDown = (e: KeyboardEvent) => {
+      if (e.key == "/") {
+        openSearch();
+        e.preventDefault();
+      }
+    };
+
+    const onCopy = (e: ClipboardEvent) => {
+      // If the user is trying to copy text normally, don't prevent them from doing so
+      const selection = window.getSelection();
+      if (selection && selection.type == "Range") return;
+
+      // Need to use window.location.hash, not just location.hash, so that the closure captures
+      // window (and thus location.hash is updated) rather than capturing location (where hash would
+      // not be updated)
+      // Avoids a dependency on points which would cause the effect to re-run often
+      const points = getHashPoints(window.location.hash, [0]);
+
+      e.clipboardData?.setData("text/plain", pointsToString(points));
+      e.preventDefault();
+    };
+
+    const onPaste = (e: ClipboardEvent) => {
+      // clipboardData may be null if the clipboard is empty, text may be empty: don't process
+      // either case
+      const text = e.clipboardData?.getData("text/plain");
+      if (!text) return;
+
+      const firstBreak = getNextClusterBreak(data, text, null);
+      if (!firstBreak) return;
+      const startUnitIndex = firstBreak.startUnitIndex;
+
+      const secondBreak = getNextClusterBreak(data, text, firstBreak);
+      if (!secondBreak) return;
+      const endUnitIndex = secondBreak.startUnitIndex;
+
+      const thirdBreak = getNextClusterBreak(data, text, secondBreak);
+      if (!thirdBreak) {
+        // Only one grapheme cluster in the pasted string
+        const cluster = text.slice(startUnitIndex, endUnitIndex);
+        // Same window vs location capture issue as in onCopy
+        window.location.hash = toFragment(stringToPoints(cluster));
+      } else {
+        // Multiple grapheme clusters in the pasted string
+        setSearchQuery(text);
+        openSearch();
+      }
+
+      e.preventDefault();
+    };
+
+    const addListeners = () => {
+      window.addEventListener("keydown", onKeyDown);
+      window.addEventListener("copy", onCopy);
+      window.addEventListener("paste", onPaste);
+    };
+    const removeListeners = () => {
+      window.removeEventListener("keydown", onKeyDown);
+      window.removeEventListener("copy", onCopy);
+      window.removeEventListener("paste", onPaste);
+    };
+
+    if (!searchOpen) {
+      addListeners();
+    } else {
+      removeListeners();
```

Comment:
if the search is open, we can just not add listeners, we don’t need to remove any listeners because the cleanup guarantees that any old listeners were removed